### PR TITLE
Initial Revision

### PR DESCRIPTION
**Pull request by Nikokrock**

### CODE REVIEW

File: e3/aws/__init__.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        else:
+            self.regions = regions
+        self.default_region = None
+        self.force_stub = stub
+        self.clients = {}
+        self.stubbers = {}
+        env = Env()
+        env.aws_env = self
+
+    def current_region(self, region=None):
```
Comment (enzbang) [inline e3/aws/__init__.py]: I would remove this function, I don't see the need.

File: e3/aws/__init__.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+        if region is None:
+            return self.default_region
+        else:
+            return region
+
+    def stub(self, name, region=None):
```
Comment (enzbang) [inline e3/aws/__init__.py]: docstring?

File: e3/aws/__init__.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+    def __enter__(self):
+        Env().aws_env.default_region = self.default_region
+
+    def __exit__(self, _type, _value, _tb):
+        del _type, _value, _tb
+        Env().aws_env.default_region = self.previous_region
+
+
+def client(name):
```
Comment (enzbang) [inline e3/aws/__init__.py]: docstring?

File: e3/aws/cfn/__init__.py [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+        :param kind: resource kind
+        :type kind: e3.aws.cfn.types.AWSType
+        """
+        assert isinstance(kind, AWSType), \
+            'resource kind should be an AWSType: found %s' % kind
+        assert name.isalnum(), \
+            'resource name should be alphanumeric: found %s' % name
+        self.name = name
+        self.kind = kind
+        self._depends = None
```
Comment (enzbang) [inline e3/aws/cfn/__init__.py]: why should it be protected?

File: e3/aws/cfn/__init__.py [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+class Stack(object):
+    """A CloudFormation stack."""
+
+    def __init__(self, name):
+        """Initialize a stack.
+
+        :param name: stack name
+        :type name: str
+        """
+        assert re.match(VALID_STACK_NAME, name) and len(name) <= 128, \
```
Comment (enzbang) [inline e3/aws/cfn/__init__.py]: can you just add a class variable to keep this value?

e.g. 


class Stack(object):

    MAX_NAME_LEN = 128

   def __init__(self, name):
      ...

File: e3/aws/cfn/__init__.py [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+
+    def export(self):
+        """Export stack as dict.
+
+        :return: a dict that can be serialized as YAML to produce a template
+        :rtype: dict
+        """
+        return {
+            'AWSTemplateFormatVersion': '2010-09-09',
+            'Description': 'no description',
```
Comment (enzbang) [inline e3/aws/cfn/__init__.py]: why?

File: e3/aws/cfn/__init__.py [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+        :param region: region in which to create the stack. If None the
+            default region is used.
+        :type region: str | None
+        """
+        return client.create_stack(StackName=self.name,
+                                   TemplateBody=self.body,
+                                   Capabilities=['CAPABILITY_IAM'])
+
+    @client('cloudformation')
+    def create_change_set(self, client, name):
```
Comment (enzbang) [inline e3/aws/cfn/__init__.py]: missing docstrings

File: e3/aws/cfn/__init__.py [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+        return client.create_change_set(ChangeSetName=name,
+                                        StackName=self.name,
+                                        TemplateBody=self.body,
+                                        Capabilities=['CAPABILITY_IAM'])
+
+    @client('cloudformation')
+    def delete(self, client):
```
Comment (enzbang) [inline e3/aws/cfn/__init__.py]: missing docstrings

File: e3/aws/cfn/__init__.py [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+        return client.delete_stack(StackName=self.name)
+
+    @client('cloudformation')
+    def cost(self, client):
```
Comment (enzbang) [inline e3/aws/cfn/__init__.py]: missing docstrings

File: e3/aws/cfn/__init__.py [hunk 10]
```diff
@@ -0,0 +1,271 @@
...
+        return client.estimate_template_cost(TemplateBody=self.body)
+
+    @client('cloudformation')
+    def resource_status(self, client, in_progress_only=True):
```
Comment (enzbang) [inline e3/aws/cfn/__init__.py]: missing docstrings

File: e3/aws/cfn/ec2/__init__.py [hunk 11]
```diff
@@ -0,0 +1,359 @@
...
+
+            self.network_interfaces[index] = device
+        elif isinstance(device, BlockDevice):
+            self.block_devices.append(device)
+        else:
+            assert False, 'invalid device %s' % device
+        return self
+
+    @property
+    def properties(self):
```
Comment (enzbang) [inline e3/aws/cfn/ec2/__init__.py]: docstring

File: e3/aws/cfn/ec2/__init__.py [hunk 12]
```diff
@@ -0,0 +1,359 @@
...
+                                          for bd in self.block_devices]}
+        if self.instance_profile is not None:
+            result['IamInstanceProfile'] = self.instance_profile.ref
+        if self.network_interfaces:
+            result['NetworkInterfaces'] = \
+                [ni.properties for ni in self.network_interfaces.values()]
+        return result
+
+
+class VPC(Resource):
```
Comment (enzbang) [inline e3/aws/cfn/ec2/__init__.py]: docstrings

File: e3/aws/cfn/ec2/__init__.py [hunk 13]
```diff
@@ -0,0 +1,359 @@
...
+    @property
+    def properties(self):
+        return {'CidrBlock': self.cidr_block}
+
+    @property
+    def cidrblock(self):
+        return self.getatt('CidrBlock')
+
+
+class Subnet(Resource):
```
Comment (enzbang) [inline e3/aws/cfn/ec2/__init__.py]: docstrings

File: e3/aws/cfn/ec2/__init__.py [hunk 14]
```diff
@@ -0,0 +1,359 @@
...
+        assert isinstance(vpc, VPC)
+        self.vpc = vpc
+
+    @property
+    def properties(self):
+        return {'CidrBlock': self.cidr_block,
+                'VpcId': self.vpc.ref}
+
+
+class InternetGateway(Resource):
```
Comment (enzbang) [inline e3/aws/cfn/ec2/__init__.py]: docstrings

File: e3/aws/cfn/ec2/__init__.py [hunk 15]
```diff
@@ -0,0 +1,359 @@
...
+        """Initialize an internet gateway.
+
+        :param name: logical name in stack
+        :type name: str
+        """
+        super(InternetGateway, self).__init__(
+            name, kind=AWSType.EC2_INTERNET_GATEWAY)
+
+
+class VPCGatewayAttachment(Resource):
```
Comment (enzbang) [inline e3/aws/cfn/ec2/__init__.py]: docstrings

File: e3/aws/cfn/ec2/__init__.py [hunk 16]
```diff
@@ -0,0 +1,359 @@
...
+
+    @property
+    def properties(self):
+        result = {'VpcId': self.vpc.ref}
+        if self.tags is not None:
+            result['Tags'] = self.tags
+        return result
+
+
+class Route(Resource):
```
Comment (enzbang) [inline e3/aws/cfn/ec2/__init__.py]: docstrings

File: e3/aws/cfn/ec2/__init__.py [hunk 17]
```diff
@@ -0,0 +1,359 @@
...
+        return self.gateway_attach.name
+
+    @property
+    def properties(self):
+        return {'RouteTableId': self.route_table.ref,
+                'DestinationCidrBlock': self.dest_cidr_block,
+                'GatewayId': self.gateway.ref}
+
+
+class SubnetRouteTableAssociation(Resource):
```
Comment (enzbang) [inline e3/aws/cfn/ec2/__init__.py]: docstrings

File: e3/aws/cfn/ec2/__init__.py [hunk 18]
```diff
@@ -0,0 +1,359 @@
...
+    def __init__(self, name, subnet, route_table):
+        """Initialize an association between a route table and a subnet.
+
+        :param name: logical name in stack
+        :type name: str
+        :param vpc: a VPC instance to attach the route table to.
```
Comment (enzbang) [inline e3/aws/cfn/ec2/__init__.py]: no vpc param but a subnet

File: e3/aws/cfn/ec2/security.py [hunk 19]
```diff
@@ -0,0 +1,150 @@
...
+                 to_port=None,
+                 description=None):
+        """Initialize a security rule.
+
+        :param target: correspond to destination (Egress rules) or source
+            (Ingress rule). The nature of the object is determined by the
+            class used. See documenation for each subclass of
+            GroupSecurityRule.
+        :type target: T
+        :param protocol: either an ip protocol name or int (see AWS
```
Comment (enzbang) [inline e3/aws/cfn/ec2/security.py]: please change the order here, protocol should be before target

File: e3/aws/ec2/ami.py [hunk 20]
```diff
@@ -0,0 +1,55 @@
...
+        :type ami_id: str
+        :param region: region in which the AMI is present. If None then
+            use default region
+        :type region: None | str
+        :param data: a dict representing the metadata of the AMI. If None then
+            download AMI description using EC2 api
+        :type data: dict | None
+        """
+        self.ami_id = ami_id
+        self.region = Env().aws_env.current_region(region)
```
Comment (enzbang) [inline e3/aws/ec2/ami.py]: no need for current_region here, just

self.region = region if region is not None else Env().aws_env.default_region

File: e3/aws/__init__.py (hunk 1 above)
Comment (Nikokrock) [reply e3/aws/__init__.py]: Done

**Pull request merged**